### PR TITLE
[Group] Decryption optimization

### PR DESCRIPTION
**Pull request by jepenven-silabs**
#### Problem
For now we decrypt with all possible keys matching the hash provided in the packet header. 
We can further decrease the number of iteration by also checking for a match with the groupId provided in the same packet header.

#### Change overview
Add a check for the groupId prior to decryption attempt. 
#### Testing
Unit test
Test suite

### CODE REVIEW

**Review by tcarmelveilleux — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/transport/SessionManager.cpp [hunk 1]
```diff
@@ -601,13 +601,16 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
...
     Credentials::GroupDataProvider * groups              = Credentials::GetGroupDataProvider();
+    GroupId groupId;
     VerifyOrReturn(nullptr != groups);
 
     if (!packetHeader.GetDestinationGroupId().HasValue())
     {
         return; // malformed packet
     }
 
+    groupId = packetHeader.GetDestinationGroupId().Value();
```
Comment (bzbarsky-apple) [inline src/transport/SessionManager.cpp:612]: Why not declare it here?  This is not C, and we are not using `exit:` stuff in this function.

File: src/transport/SessionManager.cpp [hunk 2]
```diff
@@ -630,6 +633,11 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
     bool decrypted = false;
     while (!decrypted && iter->Next(groupContext))
     {
+        // Optimization to reduce number of decryption attempt
```
Comment (bzbarsky-apple) [inline src/transport/SessionManager.cpp:636]:
```suggestion
        // Optimization to reduce number of decryption attempts
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/transport/SessionManager.cpp [hunk 3]
```diff
@@ -630,6 +633,11 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
...
+        if (groupId != groupContext.group_id)
```
Comment (msandstedt) [inline src/transport/SessionManager.cpp:637]: If we always want to additionally filter on this, why don't we add that matching logic to the iterator initialization:

```
groups->IterateGroupSessions(packetHeader.GetSessionId());
```

File: src/transport/SessionManager.cpp (hunk 1 above)
Comment (jepenven-silabs) [reply src/transport/SessionManager.cpp:612]: Done

File: src/transport/SessionManager.cpp (hunk 3 above)
Comment (jepenven-silabs) [reply src/transport/SessionManager.cpp:637]: For ease of implementation and we might not always have the groupId.

**Pull request merged**